combat-trainer.lic

```diff
@@ -3511,18 +3552,64 @@ class GameState
       return false
     end
 
-    find_cambrinth(@cambrinth, @stored_cambrinth, @cambrinth_cap)
-
     next_charge = @charges.pop
     echo("check_charging?: #{next_charge}") if $debug_mode_ct
 
+    next_charge = avtalia_charge(next_charge)
+    return true if next_charge.zero?
+    find_cambrinth(@cambrinth, @stored_cambrinth, @cambrinth_cap)
     @charges = [] unless charge?(@cambrinth, next_charge)
 
     stow_cambrinth(@cambrinth, @stored_cambrinth, @cambrinth_cap)
 
     true
   end
 
+  def avtalia_charge(charge_num) # sources mana from filled avtalia cambrinth, then returns the amount left uncharged
+    return charge_num unless DRStats.trader?
+    return charge_num unless @cambrinth_invoke_exact_amount
+    return charge_num if @avtalia_array.empty?
+    return charge_num unless DRStats.mana < 80
+
+    remaining_charge = @charges.inject(0, :+) + charge_num
+    @avtalia_cambrinth = find_avtalia(remaining_charge) if @avtalia_cambrinth.empty?
```

Comment:
❓ Why is `@avtalia_cambrinth` stored in an instance var? Couldn't it change each time we need to charge? And it's cleared out (to `{}`) most of the time.

Reply:
Yep,  A relic from when I had all the common logic in combat-trainer.  I thought it'd be sitting around doing more in gamestate, but it turned out not to be necessary.

Reply:
I remembered why it's stored as an instance variable.  avtalia_charge is a little bit recursive as it unwraps bites of the @charges arrays, and I keep it instanced so that I can avoid calling DRCA.choose_avtalia if avtalia_cambrinth is already set and potentially changing ships mid-stream.  See:
```
    # unwrap the charges for the spell. Keep pulling in more cambrinth mana entries until it either hits the camb cap or gets all of it.
    return avtalia_charge(charge_num + @charges.pop) if charge_num < [@avtalia_cambrinth.last['mana'], remaining_charge].min
```

I can still do it another way, but there's value to this if I want anything else to be aware that I've got a lien on that piece of jewelry while I'm prepping to invoke it.  I need to think about that to see if that's a real situation I'd ever run into, or just fantasy.